src/compiler/checker.ts

```diff
@@ -12118,7 +12118,7 @@ namespace ts {
                         }
                     }
                 }
-                return getIntersectionType(constraints);
+                return getNormalizedType(getIntersectionType(constraints), /*writing*/ false); // The source types were normalized; ensure the result is normalized too.

```

Comment:
the `writing` parameter has me a little suspicious here.

Side note, the comment makes this line a bit too long.

```suggestion
                // The source types were normalized; ensure the result is normalized too.
                return getNormalizedType(getIntersectionType(constraints), /*writing*/ false); 
```

Reply:
This function is only called on the source types in relation; this mirrors how the same function calls `getNormalizedType` on the source itself earlier. I had this instead at the one call site, but opted to stick it here because there is only one call site and it's shorter than:

```ts
let constraint = getEffectiveConstraintOfIntersection(source.flags & TypeFlags.Intersection ? (source as IntersectionType).types: [source], !!(target.flags & TypeFlags.Union));
if (constraint) {
    constraint = getNormalizedType(constraint, /*writing*/ false);
}
if (constraint && everyType(constraint, c => c !== source)) { // Skip comparison if expansion contains the source itself
    // ...
}
```

But, I am happy to do so.

Reply:
The other option is maybe to move all of those argument expressions into `getEffectiveConstraintOfIntersection` like

```ts
function getEffectiveConstraintOfIntersection(source: Type, target: Type) {
    const types = source.flags & TypeFlags.Intersection ? (source as IntersectionType).types: [source];
    const targetIsUnion = !!(target.flags & TypeFlags.Union);
    // ...
}
```

And then move it into the relation closure to make it obvious that this should only be used in this one place and not for another purpose?

Reply:
FWIW the above is sort of what I'd prefer (accept `source` and `target`). Better yet, move `getEffectiveConstraintOfIntersection` down into relation to make it obvious what you are/aren't supposed to use it for, but the diff won't be followable.

Reply:
I've moved the comment per Daniel's comment.